New article: Getting started with the JavaScript SDKs (#29)
sharetribe/dev-docs: content/introduction/getting-started-with-sdks/index.mdx

      <Table.Th>Sharetribe SDK</Table.Th>
      <Table.Th>Sharetribe Integration SDK</Table.Th>
    </Table.Tr>
  </thead>
  <tbody>
    <Table.Tr>
      <Table.Td>Intended Use</Table.Td>
      <Table.Td>
        Most common API requests with Marketplace API or Asset Delivery
        API
      </Table.Td>
      <Table.Td>Operator-level requests with Integration API</Table.Td>
    </Table.Tr>
    <Table.Tr>
      <Table.Td>Required Parameters</Table.Td>
      <Table.Td>Client ID</Table.Td>
      <Table.Td>Client ID, Client secret</Table.Td>
    </Table.Tr>
    <Table.Tr>
      <Table.Td>Secure context needed</Table.Td>
      <Table.Td>No</Table.Td>
      <Table.Td>**Yes**</Table.Td>
    </Table.Tr>
  </tbody>
</Table>

To try out the Sharetribe SDK, visit our
[Github documentation](https://sharetribe.GitHub.io/flex-sdk-js/index.html)
for instructions on testing it directly in your browser or via the
[API Playground](https://sharetribe.GitHub.io/flex-sdk-js/try-it-in-the-playground.html).
For a more detailed guide on using the SDK with code, keep following
this tutorial.

## Installation with Yarn Package Manager

To get started with the SDK, you'll first need to install

> [Yarn Package Manager](https://yarnpkg.com/). After that is set up, you

Reviewer: Would it make sense to offer both yarn and npm as options here? Since folks are using their own clients, they might already be using npm and not want to install yarn for just this use case.

can run the following command in your project directory:

```jsx
yarn add sharetribe-flex-sdk
```

<Callout type="info">
  The above example will create an instance that allows you to access
  Marketplace API and Asset Delivery API. In order to access Integration
  API, run `yarn add sharetribe-flex-integration-sdk`
</Callout>

Ensure you have created an application in Console. If you have not
already done this when setting up the Sharetribe Web Template, you can
follow the instructions in

> [this tutorial](https://www.sharetribe.com/docs/introduction/getting-started-with-web-template/#sharetribe-client-id-and-client-secret)

Reviewer: I think we could have the same application creating instructions here, too – that way, folks don't need to click around and then remember to come back to this article (especially if there's no return link in the instructions). We can reuse the same images here, so no need to create anything new. It's a bit of duplication, but I would say it's worth it for the nicer user experience.

```jsx
const sharetribeSdk = require('sharetribe-flex-sdk');

const sdk = sharetribeSdk.createInstance({
  clientId: '<Client ID>',
});
```

If you are accessing Integration API, you'll need to also provide the
client secret.

```jsx
const sharetribeIntegrationSdk = require('sharetribe-flex-integration-sdk');

const integrationSdk = sharetribeIntegrationSdk.createInstance({
  clientId: '<Client ID>',
  clientSecret: '<Client secret>',
});
```

This will allow you to make requests that don't require authentication
or a trusted SDK instance, e.g. querying listings. We'll cover how to
create a trusted SDK in the next step.

## Creating a trusted SDK instance

To make certain requests, you will need to access the APIs from a
trusted context. In order to do this, you'll need to log in using the
SDK and perform a token exchange to retrieve a trusted token.

1. Create an instance of the SDK with the client ID, secret, and token
   store

```jsx
const sharetribeSdk = require('sharetribe-flex-sdk');

var sdk = sharetribeSdk.createInstance({
  clientId: '<Client ID>',
  clientSecret: '<Client secret>',
  tokenStore: sharetribeSdk.tokenStore.memoryStore(),
});
```

<Callout type="info">

>   You can follow the same process above when working with Integration

Author: @SariSaar I meant that you can use the same method to create an instance of the integration SDK - perhaps this is better as a comment though

  API, just remember to import the correct SDK with the `require`
  keyword.
</Callout>

2. Log in using valid credentials

```jsx
sdk.login({ username: '<User email>', password: '<User password>' });
```

3. Exchange the access token for a trusted token

```jsx
sdk.exchangeToken().then((res) => {
  var store = sharetribeSdk.tokenStore.memoryStore();
  store.setToken(res.data);
});
```

4. Create an instance of the SDK using the trusted token

```jsx
var sdkTrusted = sharetribeSdk.createInstance({
  clientId: '<Client ID>',
  tokenStore: sharetribeSdk.tokenStore.memoryStore(),
});
```

Now you can use `sdkTrusted` to make requests that rely on a secure
context, e.g. deleting a user with
[Marketplace API](https://www.sharetribe.com/api-reference/marketplace.html#delete-user).

<Callout type="info">
  Requests that require a trusted SDK rely on valid login credentials
  appropriate for the request. For example, to delete a user you must
  log in with that user's credentials.
</Callout>

## Example usage

> Here are a few examples demonstrating common uses and capabilities of

Reviewer 2: One additional example could also be querying the asset delivery API. The asset delivery API is mentioned in the article and it could be a good idea to demonstrate that the Sharetribe SDK is the right SDK to call it

the SDKs.

### Logging in

```jsx
const sharetribeSdk = require('sharetribe-flex-sdk');

var sdk = sharetribeSdk.createInstance({
  clientId: '<Client ID>',
});

sdk
  .login({ username: '<User email>', password: '<User password>' })
  .then((_) => {

>     // Request dependent on a trusted access token

Reviewer: This is the term the API reference uses for this level of authentication

  });
```

### Initializing a trusted SDK

> 

Reviewer 2: Hijacking this comment thread but @SariSaar @Amanda-Kauppinen, I'm wondering if "Trusted SDK" is an accurate term to use – it kind of implies there would be a third JavaScript SDK. There's some ambiguity whether trusted SDK refers to another SDK, or an SDK  instance initialised with a trusted token.

Do you think that's a necessary distinction to make? An alternative would be to use e.g. "SDK initialised with trusted access token" and shorten it to "SDK (trusted token)"

```jsx
const sharetribeSdk = require('sharetribe-flex-sdk');

var sdk = sharetribeSdk.createInstance({
  clientId: '<Client ID>',
  clientSecret: '<Client secret>',
  tokenStore: sharetribeSdk.tokenStore.memoryStore(),
});

sdk
  .login({ username: '<User email>', password: '<User password>' })
  .then((_) => {
    sdk.exchangeToken().then((res) => {
      var store = sharetribeSdk.tokenStore.memoryStore();
      store.setToken(res.data);

      var sdkTrusted = sharetribeSdk.createInstance({
        clientId: '<Client ID>',
        tokenStore: store,
      });

      // Place the request requiring trusted access here
    });
  });
```

### Updating a listing

```jsx
const sharetribeSdk = require('sharetribe-flex-sdk');

var sdk = sharetribeSdk.createInstance({
  clientId: '<Client ID>',
});

sdk
  .login({ username: '<User email>', password: '<User password>' })
  .then((_) => {
    sdk.ownListings
      .update({
        id: '<Listing ID>',
      })
      .then((res) => {
        // res.data
      });
  });
```

<Callout>

>   Updating a listing requires that the listing author's credentials are

Reviewer: If you add "intros" to each use case, this detail could work as that intro text

  provided in the log in step
</Callout>

### Updating metadata in a transactions

```jsx
const IntegrationSdk = require('sharetribe-flex-integration-sdk');

var sdk = sharetribeSdk.createInstance({
  clientId: '<Client ID>',
  clientSecret: '<Client secret>',
});

sdk.transactions
  .updateMetadata({
    id: '<Transaction ID>',
    metadata: {
      '<Key>': '<Value>',
    },
  })
  .then((res) => {
    // res.data
  });
```

## Additional resources

For the complete documentation on Sharetribe’s SDKs, visit our
[GitHub documentation](https://sharetribe.github.io/flex-sdk-js/index.html).
There, you'll find:

- Features
- Configuration options
- Additional examples
- How to use the browser or the
  [API Playground](https://sharetribe.GitHub.io/flex-sdk-js/try-it-in-the-playground.html)
  for easy testing

<Callout type="info">
  If you want to get experience testing the Integration SDK, you will
  need to access it from a secure context. An easy way to do this is to

>   run a small server from which to make the requests.

Reviewer: It looks like this maybe wasn't resolved yet?

>   run a small server from which to make the requests.

Author: @SariSaar I did [rework this section](https://github.com/sharetribe/dev-docs/blob/29be6f7a8334a6c2e1b5138a76950394f4c5cfe6/content/introduction/getting-started-with-sdks/index.mdx?plain=1#L465), but kept the callout to indicate that  secure context is sometimes needed.

> For the complete documentation on Sharetribe's SDKs, visit our Github documentation for the [Standard SDK](https://sharetribe.github.io/flex-sdk-js/index.html) and the [Integration SDK](https://sharetribe.github.io/flex-integration-sdk-js/). There, you'll find
> 
>     Features
>     Configuration options
>     Additional examples
>     How to use the browser or the [API Playground](https://sharetribe.github.io/flex-sdk-js/try-it-in-the-playground.html) (only available for the standard SDK) for easy testing
> 
> Certain requests will require a secure environment setup. An easy way to do this is to run a small server from which to make these types of requests.

Do you think it needs to be changed?

>   run a small server from which to make the requests.

Author: Ah great comment!

<br />
<Table>
  <thead>
    <Table.Tr>
      <Table.Th />
      <Table.Th>Sharetribe SDK</Table.Th>
      <Table.Th>Sharetribe Integration SDK</Table.Th>
    </Table.Tr>
  </thead>
  <tbody>
    <Table.Tr>
      <Table.Td>Intended Use</Table.Td>
      <Table.Td>
        Most common API requests with Marketplace API or Asset Delivery
        API
      </Table.Td>
      <Table.Td>Operator-level requests with Integration API</Table.Td>
    </Table.Tr>
    <Table.Tr>
      <Table.Td>Required Parameters</Table.Td>
      <Table.Td>Client ID</Table.Td>
      <Table.Td>Client ID, Client secret</Table.Td>
    </Table.Tr>
    <Table.Tr>
      <Table.Td>Secure context needed</Table.Td>
      <Table.Td>No</Table.Td>
      <Table.Td>**Yes**</Table.Td>
    </Table.Tr>
  </tbody>
</Table>

To try out the Sharetribe SDK, visit our
[Github documentation](https://sharetribe.GitHub.io/flex-sdk-js/index.html)
for instructions on testing it directly in your browser or via the
[API Playground](https://sharetribe.GitHub.io/flex-sdk-js/try-it-in-the-playground.html).
For a more detailed guide on using the SDK with code, keep following
this tutorial.

> ## Installation

Author: @shareoc yeah I had tried this and then scrapped it when I realised you can't do the code block annotation - ```. I thought it looked a bit odd with the inline syntax, but happy to use tabs if I'm being over critical 😄

To get started with the SDK, you'll first need to install it with either
[Yarn](https://yarnpkg.com/)

```jsx
yarn add sharetribe-flex-sdk
```

or [npm](https://www.npmjs.com/)

```jsx
npm install sharetribe-flex-sdk
```

<Callout type="info">
  The above example will create an instance that allows you to access
  Marketplace API and Asset Delivery API. In order to access Integration
  API, run `yarn add sharetribe-flex-integration-sdk`
</Callout>

## SDK Initialization Overview

### API applications

To use the Marketplace API or Integration API, you will need a **client
ID**.

When you get access, you will be able to log into Sharetribe Console and
check the client ID.<br /> Sharetribe Console: _Build > Advanced >
Applications_

In addition, Sharetribe Web Template uses transaction processes that
include privileged transitions. This makes it possible to customize
pricing on the Node server that's included in the template. The **client
secret** is needed to make this secure call from the template's own
server to Sharetribe API.

> 

Reviewer 2: My suggestion would not be to mention the template here as the primary example. Privileged transitions can be used regardless of template usage or not. Technically, the client secret is needed to fetch the Trusted user access token, which is needed to call certain endpoints, including the transition transaction endpoint when transitioning a privileged transition. These calls should be made from a secure context. You could explain that and then illustrate how that works in practice by using the template as an example?

- Asset Delivery API, for delivering assets configured in Console
- Integration API, for advanced operational use

You can find more information on API usage and guidelines for choosing
which to work with in
[this article](https://www.sharetribe.com/docs/concepts/api-sdk/marketplace-api-integration-api/#when-to-use-the-marketplace-api).

There are two SDK variants in which to access the APIs

<br />
<Table>
  <thead>
    <Table.Tr>
      <Table.Th />
      <Table.Th>Sharetribe SDK</Table.Th>
      <Table.Th>Sharetribe Integration SDK</Table.Th>
    </Table.Tr>
  </thead>
  <tbody>
    <Table.Tr>
      <Table.Td>**Access level**</Table.Td>
      <Table.Td>Limited to user-scope and public data</Table.Td>
      <Table.Td>Full marketplace (operator-level) access</Table.Td>
    </Table.Tr>
    <Table.Tr>
      <Table.Td>**API access**</Table.Td>
      <Table.Td>Marketplace API and Asset Delivery API</Table.Td>
      <Table.Td>Integration API</Table.Td>
    </Table.Tr>
    <Table.Tr>
      <Table.Td>**Required Parameters**</Table.Td>
      <Table.Td>
        Client ID (Client secret for trusted contexts)
      </Table.Td>
      <Table.Td>

>         Client ID (Client secret for trusted contexts)

Reviewer: For Integration API, client secret and secure context are always needed, so no need for the mention of trusted context/SDK on that side.

- The request is made from a secure context.

<Callout type="info">
  In the [Sharetribe Web
  Template](https://github.com/sharetribe/web-template), these types of
  requests are made on an small Node server included in the Template.
</Callout>

### Initializing a basic SDK instance

You can initialize a basic SDK instance using only the application's
Client ID. Ensure the client ID is from a **Marketplace API
application** created in Console.

```jsx
const sharetribeSdk = require('sharetribe-flex-sdk');

const sdk = sharetribeSdk.createInstance({
  clientId: '<Client ID>',
});
```

This will allow you to make requests that don't require authentication
or a trusted SDK instance, e.g. querying listings.

### Initializing an Integration SDK instance

If you are accessing Integration API, you'll need to also provide the
client secret. Ensure the client ID is from an **Integration API
application** created in Console.

```jsx
const sharetribeIntegrationSdk = require('sharetribe-flex-integration-sdk');

const integrationSdk = sharetribeIntegrationSdk.createInstance({
  clientId: '<Client ID>',
  clientSecret: '<Client secret>',
});
```

### Initializing a trusted SDK instance

> To make certain requests, you will need to access the APIs from a

Reviewer: ```suggestion
To make certain requests, you will need to access the Marketplace API from a
```

trusted context. In order to do this, you'll need to log in using the
SDK and perform a token exchange to retrieve a trusted token.

> <Steps>

Reviewer 2: I forgot to mention, but there are two ways of using the Steps component. Using markdown headings (i.e. `###`) is one way, and that way these titles will also show up in the TOC.

An alternative is to use the following syntax:

`{<h3>Create an instance of the SDK with the client ID, secret, and token store</h3>}`

This results in the same outcome, except that the titles will not be shown in the TOC. 

I think in this case it makes sense to use the latter, as semantically, these are steps and not headings. Sorry, should've mentioned that earlier!

## Create an instance of the SDK with the client ID, secret, and token store

```jsx

> // use 'sharetribe-flex-integration-sdk' here to create and integration SDK instance

Reviewer: This comment can be removed, because the trusted exchange only works for Marketplace API (Integration API is a trusted context by default and doesn't have user logins)

  .then((_) => {
    // Request dependent on an authenticated user access token
  });
```

### Initiating a transaction

**API**: Marketplace API

The process of initiating a transaction is highly variable because it
depends entirely on transaction flow customizations and the required
paramters of the transition.

For this example, we'll use the default-booking process, and choose
`transition/request-payment` as the first step. This requires that:

- The customer is logged in
- A token exchange is made to get a trusted user access token
- The request includes the following parameters
  - `transition`
  - `processAlias`
  - `params`
    - `listingID`
    - `bookingStart`
    - `bookingEnd`
    - `lineItems`

<Callout type="info">
  This example uses the **minimum** required parameters per the
  transition defined in the transaction process. You'll likely want to
  include more information during this step, e.g. updating the
  transaction's extended data. See the [Marketplace API reference
  documentation](https://www.sharetribe.com/api-reference/marketplace.html#initiate-transaction)
  for more information.
</Callout>

> ```jsx

Reviewer 2: Could add SDK import to the beginning here to be consistent with the other examples, i.e.
`const sharetribeSdk = require('sharetribe-flex-sdk');`

const transition = 'transition/request-payment';
const processAlias = 'default-booking/release-1';

const params = {
  listingId: '<Listing ID>',
  bookingStart: '2025-12-20T22:00:00.000Z',
  bookingEnd: '2025-12-21T22:00:00.000Z',
  lineItems: [
    {
      code: 'line-item/night',
      unitPrice: {
        _sdkType: 'Money',
        amount: 5000,
        currency: 'EUR',
      },
      quantity: 1,
      includeFor: ['customer', 'provider'],
    },
    {
      code: 'line-item/provider-commission',
      unitPrice: {
        _sdkType: 'Money',
        amount: 5000,
        currency: 'EUR',
      },
      percentage: -13,
      includeFor: ['provider'],
    },
  ],
};

const sdk = sharetribeSdk.createInstance({

>   clientID: '<Client ID>',

Reviewer 2: ```suggestion
  clientId: '<Client ID>',
```

  clientSecret: '<Client secret>',

>   tokenStore: sharetribeSdk.tokenStore.memoryStore(),

Reviewer 2: I think you can leave this tokenStore out when you are just initialising the Sharetribe SDK

      var store = sharetribeSdk.tokenStore.memoryStore();
      store.setToken(res.data);

      var sdkTrusted = sharetribeSdk.createInstance({
        clientId: '<Client ID>',
        tokenStore: store,
      });

      // It's useful to first call initiateSpeculative to validate the data and to ensure
      // the real request will complete as expected
      sdkTrusted.transactions
        .initiateSpeculative(
          {
            transition,
            processAlias,
            params,
          },
          {
            expand: true,
          }
        )
        .then((speculatedResult) => {
          if (speculatedResult.status == 200) {
            sdkTrusted.transactions
              .initiate({
                transition,
                processAlias,
                params,
              })
              .then((initiateResult) => {
                // initiateResult.data
              });
          }
        });
    });
  });
```

### Transitioning a transaction

**API**: Marketplace API

This example shows how you can transition a transaction that has already
been initiated. The logged-in user must be a party that is allowed to

> perform the transition - either operator, customer, or provider. This

Reviewer: Thinking out loud here – not all transitions require a trusted context, but some of them do. What do you think, would it make sense to mention in this description / add a callout that for non-privileged transitions, a login is enough (and then link to the login example) and for privileged transitions, folks will need to use the trusted SDK? We could link then to the privileged transitions Docs article for context so we don't need to explain it here.

can be checked in the transaction process visualizer in Console, or from
the transaction's `process.edn` file.

```jsx
const sharetribeSdk = require('sharetribe-flex-sdk');

var sdk = sharetribeSdk.createInstance({
  clientId: '<Client ID>',
  clientSecret: '<Client secret>',
  tokenStore: sharetribeSdk.tokenStore.memoryStore(),
});

sdk
  .login({ username: '<User email>', password: '<User password>' })
  .then((_) => {
    sdk.exchangeToken().then((res) => {
      var store = sharetribeSdk.tokenStore.memoryStore();
      store.setToken(res.data);

      var sdkTrusted = sharetribeSdk.createInstance({

>         clientId: '<Client secret>',

Reviewer 2: ```suggestion
        clientId: '<Client ID>',
```

        tokenStore: store,
      });

      sdkTrusted.transactions
        .transition({
          id: '<Transaction ID>',
          transition: '<Transition name>',
          params: {},
        })
        .then((res) => {
          // res.data
        });
    });
  });
```

### Updating metadata in a transaction

**API**: Integration API

Updating metadata for a transaction requires that you use the
Integration API.

```jsx
const IntegrationSdk = require('sharetribe-flex-integration-sdk');

> var sdk = sharetribeSdk.createInstance({

Reviewer 2: ```suggestion
var sdk = IntegrationSdk.createInstance({
```

  clientId: '<Client ID>',
  clientSecret: '<Client secret>',
});

sdk.transactions
  .updateMetadata({
    id: '<Transaction ID>',
    metadata: {
      '<Key>': '<Value>',
    },
  })
  .then((res) => {
    // res.data
  });
```

### Updating a listing

**API**: Marketplace API

Updating a listing requires that the listing author's credentials are
used.

```jsx
const sharetribeSdk = require('sharetribe-flex-sdk');

var sdk = sharetribeSdk.createInstance({
  clientId: '<Client ID>',
});

sdk
  .login({ username: '<User email>', password: '<User password>' })
  .then((_) => {
    sdk.ownListings
      .update({
        id: '<Listing ID>',

>         description: '<Updated description',

Reviewer: ```suggestion
        description: '<Updated description>',
```

        privateData: {
          '<Key>': '<Value>',
        },
        publicData: {
          '<Key>': '<Value>',
        },
      })
      .then((res) => {
        // res.data
      });
  });
```

## Additional resources

For the complete documentation on Sharetribe's SDKs, visit our Github
documentation for the
[Standard SDK](https://sharetribe.github.io/flex-sdk-js/index.html) and
the
[Integration SDK](https://sharetribe.github.io/flex-integration-sdk-js/).
There, you'll find

- Features
- Configuration options
- Additional examples
- How to use the browser or the
  [API Playground](https://sharetribe.GitHub.io/flex-sdk-js/try-it-in-the-playground.html)
  (only available for the standard SDK) for easy testing

> <Callout type="info">

Reviewer 2: Might be worth accompanying this with a "Never expose the client secret in frontend code." warning callout

---
title: JavaScript SDKs
description:
  This article serves as a guide on how to get started using
  Sharetribe's JavaScript SDKs.
---

import { Callout, Table, Steps, Tabs } from 'nextra/components';

# Getting started with the JavaScript SDKs

Sharetribe's
[JavaScript Software Development Kit](https://sharetribe.github.io/flex-sdk-js/)
(SDK) and
[Integration SDK](https://sharetribe.github.io/flex-integration-sdk-js/)
simplify interactions with our APIs. You can read more about features
and functionalities in our
[SDK GitHub Documentation](https://sharetribe.GitHub.io/flex-sdk-js/index.html).
This guide covers

- Installation
- Initialization and instantiating the SDK
- Creating a trusted SDK instance
- Example use cases

> ## Overview

Reviewer 3: Comment/Change request:

In this document and especially in this paragraph, we use the terms related to the secure context a bit loosely.

### "trusted context" vs "secure context"

In the article, both terms are used interchangeably. I think we should pick one and use it consistenty and clearly define it in laymans terms.

The term that is used in API reference is **secure context** (https://www.sharetribe.com/api-reference/authentication.html).

In the docs, I noticed that we also use "trusted context". This is not terribly wrong in my opinion, but it's something we've used mainly inside the core codebase, but not so much externally.

Also, the definition of secure context:

* Server that only the operator has access to -> trusted
* Publicly accessible website where the source code is visible to public (i.e. any website basically) -> untrusted
* Password protected website where only the operator has access to -> trusted
* Mobile app where the client secrets might leak to the public -> untrusted

### "trusted requests"

I'm not sure we have used this term before, but conceptually it makes sense. Requests made with trusted user access token -> "trusted requests"

> ## Overview

Reviewer 3: Read this comment first!

In this paragraph and in this document in general, we use some terms a bit loosely. It's not a surprise, I think we haven't clearly defined them anywhere before this.

I'll go first through all the steps that are needed in order to do requests like marketplace delete and I'll write down the relationship between these steps:

1. Some requests (like deleting user) require the request to be made with a **trusted user access token[1]**. We need a **SDK instance holding the trusted user access token[2]** to do those
2. To initialize **SDK instance holding the trusted user access token** we need to have the trusted user access token, which we can get by doing a token exchange
3. Token exchange takes **Client Secret** + user access token of an authenticated users (i.e. logged in user) and returns **trusted user access token**. In order to token exchange, we need an SDK instance initialized with **Client Secret**
4. **Client Secret** is a secret that must not be leaked to the public. That's why, we can only use **Client Secret** and initialize SDK with it in a **secure context[3]**

So these steps have clear dependencies:

requests with trusted user access token 
--> REQUIRES SDK instance with trusted access token 
  --> REQUIRES trusted access token obtained by token exchange which
    --> REQUIRES SDK instance initialized with Client Secret 
       --> REQUIRES secure context because the Client Secret must not leak

In the article, I think, it's ok to skip some steps. For example, I think it is ok to say something like "deleting a user requires secure context", because this is implied. _Technically_, this isn't 100% accurate, because the only thing the API cares about is that it gets a trusted user access token. However, it's implied that in order to get the trusted user access token, you should use Client Secret, which you should only use in a secure context.

But saying something that goes to the opposite direction is incorrect. For example

> Some requests must be made from a _trusted context_, meaning
the SDK must be initialized with both the client ID and client secret.
As a result, all Integration SDK requests are performed within a trusted
context.

The above is wrong, because initializing SDK with Client Secret doesn't make the context trusted. If you initialize Integration SDK in a web browser in a website that is publicly available and you use Client Secret and leak it to the public, it doesn't make the context trusted.

Next, I'll list the terms and the above steps and discuss a bit about them

### [1] Requests that require trusted user access token / "trusted requests"

"trusted user access token" is clearly defined term in Authentication API reference (https://www.sharetribe.com/api-reference/authentication.html) and it's correctly used in this article.

"trusted requests" is something I'm not sure we have used before. If you want to be explicit, you can say "requests that require trusted user access token". It's long, but it's explicit. "trusted requests" is in my opinion understandable though.

### [2] SDK instance holding the trusted user access token / "trusted SDK instance"

"trusted SDK instance" is a term I'm not sure we have used before. I think it makes conceptually sense, but if you want to be more explicit, you can say "SDK instance holding the trusted user access token".

### [3] Secure context / "trusted context"

The article uses both terms interchangeably. It's not wrong IMO, they mean practically the same (I can't come up with an example where context would be secure but not trusted, or vice versa).

I'd still recommend picking one and using it consistently in the article. The one I'd recommend is **Secure context**.

Secure context is a term we use in Authentication API reference (https://www.sharetribe.com/api-reference/authentication.html).

"Trusted context" is something I can see used quite widely in the Dev Docs. This is a term we use internally in core codebase for certain cases, but not outside of code. I believe it was "leaked" from code codebase to Dev Docs :)

Anyway, as said, "trusted context" isn't wrong. I think you can pick it if you want to be consistent with the rest of the docs in Dev Docs. Or change all Dev Docs pages use "secure context".

Also, "context" is a bit generic. A more descriptive and maybe human friendly and at the same time correct term would be "environment" or "runtime environment". So, "secure environment" or "secure runtime environment". I think "environment" highlights that we are talking about the environment in which the SDK is being used (website, backend server, mobile app etc.)

Reviewer 2: @Amanda-Kauppinen Does this make sense? 

My key takeaways:
- Let's stop using the terminology "secure context" and "trusted context"
- Let's define what "Trusted SDK Instance" means, i.e. -> SDK instance holding the trusted user access token (this is what previously was referred to as "trusted context"
- "Secure context" should become "secure runtime environment, e.g. server", or similar

Our SDKs simplify making requests to the backend by handling
authentication and providing convenient access to Sharetribe's APIs.
They allow you to interact easily with

- [Marketplace API](https://www.sharetribe.com/api-reference/marketplace.html),
  for general interactions with marketplace data
- [Integration API](https://www.sharetribe.com/api-reference/integration.html),
  for advanced operational use
- [Asset Delivery API](https://www.sharetribe.com/api-reference/asset-delivery-api.html),
  for delivering assets configured in Console

You can find more information on API usage and guidelines for choosing
which to work with in
[this article](https://www.sharetribe.com/docs/concepts/api-sdk/marketplace-api-integration-api/#when-to-use-the-marketplace-api).

There are two JavaScript SDKs you can use to access the APIs

<br />
<Table>
  <thead>
    <Table.Tr>
      <Table.Th />
      <Table.Th>Sharetribe SDK</Table.Th>
      <Table.Th>Sharetribe Integration SDK</Table.Th>
    </Table.Tr>
  </thead>
  <tbody>
    <Table.Tr>
      <Table.Td>**Access level**</Table.Td>

>       <Table.Td>Limited to user-scope and public data</Table.Td>

Reviewer 3: Comment: Hmhm... "public data" might be a bit confusing. To me, this sounds like I can only access publicly available listings etc. but when the user is logged in, they can also change their private data etc., things that are not "public" to the outside world

Reviewer 3: Oh nvm, now that I read it again I think it makes sense. "and" is important :) So "user-scope" + all publicly available.

      <Table.Td>Full marketplace (operator-level) access</Table.Td>
    </Table.Tr>
    <Table.Tr>
      <Table.Td>**API access**</Table.Td>
      <Table.Td>Marketplace API and Asset Delivery API</Table.Td>
      <Table.Td>Integration API</Table.Td>
    </Table.Tr>
    <Table.Tr>

>       <Table.Td>**Required Parameters**</Table.Td>

Reviewer 3: Comment: "parameters" is quite generic, many things can be parameters. I think a more descriptive word would be "Client credentials", because these are specific "parameters" that we use to authenticate the SDK

      <Table.Td>
        Client ID (Client secret for trusted contexts)
      </Table.Td>

>       <Table.Td>Client ID, client secret</Table.Td>

Reviewer 3: Comment: A bit of inconsistent capitalization, not only here but in the document in general. At least in the UI we use Client ID and Client Secret. I think we should use those.

    </Table.Tr>
    <Table.Tr>
      <Table.Td>**Secure context needed**</Table.Td>

>       <Table.Td>When initialised with a trusted access token</Table.Td>

Reviewer 3: ```suggestion
      <Table.Td>When initialised with a Client Secret</Table.Td>
```

      <Table.Td>Yes</Table.Td>
    </Table.Tr>
  </tbody>
</Table>

You'll notice we mention secure and trusted contexts in the above
comparison. Some requests must be made from a _trusted context_, meaning

> the SDK must be initialized with both the client ID and client secret.

Reviewer 3: Comment: The way I think about "trusted context" is that it's a context in which only the operator has access and ability to make SDK requests. So, for example:

* Server that only the operator has access to -> trusted
* Website where the JavaScript and HTML source code is available to the end user to see and use -> untrusted
* Password protected website where only the operator has access to -> trusted

Maybe this is splitting the hair a bit, but I think "trusted context" doesn't mean tha the SDK is initialized with client ID and client secret. It means that the SDK is used in a context that isn't publicly available. And what follows from there is that when the context is not publicly available, you can use the client secret there.

Does that make sense?

As a result, all Integration SDK requests are performed within a trusted

> context.

Reviewer 3: > Some requests must be made from a _trusted context_,

This is fine

> meaning the SDK must be initialized with both the client ID and client secret.

Instead, I'd explain what secure / trusted context is, for example: "Secure context is a runtime environment where you can initialize SDK with Client Secret without exposing it to the public."

> As a result, all Integration SDK requests are performed within a trusted
context.

This is a bit wrong. This sounds to me that using Client Secret makes the context secure, which is not the case. One can create Integration SDK with Client ID and Client Secret in a publicly available website and it doesn't make that website a secure context.

All trusted requests must be made from a _secure context_, meaning they

> need to be made from a server-side environment. An easy way to do this

Reviewer 3: ```suggestion
need to be made from a server-side environment. One way to do this
```

I would avoid using "easy". Someone less technical might be demotivated if we say that setting up a server is "easy". Easy for some for sure, not for everyone.

is to run a small server from which to make these types of requests.

To try out the SDKs, visit our
[Github documentation for the Sharetribe SDK](https://sharetribe.GitHub.io/flex-sdk-js/index.html)
and the
[Integration SDK](https://sharetribe.github.io/flex-integration-sdk-js/)
for instructions on how to get started. For a more detailed guide on
using the SDKs with code, keep following this tutorial.

<Callout type="info">

>   Certain requests will require a secure context. An easy way to do this

Reviewer 3: Same comment about "easy"

  is to run a small server from which to make these types of requests.
</Callout>

## Installation

To get started with the SDKs, you'll first need to install them with
either [Yarn](https://yarnpkg.com/) or [npm](https://www.npmjs.com/).

{/* prettier-ignore */}
<Tabs items={['yarn', 'npm']}>
  <Tabs.Tab>
  ```shell 
  yarn add sharetribe-flex-sdk
  
  yarn add sharetribe-flex-integration-sdk
  ```
  </Tabs.Tab>
  <Tabs.Tab>
  ```shell 
  npm install sharetribe-flex-sdk
  
  npm install sharetribe-flex-integration-sdk
  ```
  </Tabs.Tab>
</Tabs>
{/* // prettier-ignore */}

## SDK Initialization Overview

### Obtaining the Client ID and secret

To use the Marketplace API or Integration API, you will need a **client
ID** from an application. You can create an application in Sharetribe
Console: _Build > Advanced > Applications_

![Sharetribe Console: Applications tab](./console-build-application.png)

You should also make note of the **client secret**, as some requests
require a secure context and this can only be done when initializing the

> SDK with both the client ID and secret.

Reviewer 3: ```suggestion
SDK with both the Client ID and Client Secret and after obtaining the trusted user access token by token exchange
```

An example of this type of request is making a privileged transition in
a transaction. This requires that the:

- Client ID and secret are used when creating an instance of the SDK
- A token exchange happens to obtain a trusted user access token
- The request is made from a secure context.

<Callout type="info">
  In the [Sharetribe Web
  Template](https://github.com/sharetribe/web-template), these types of
  requests are made on a small Node server included in the Template.
</Callout>

### Initializing a Sharetribe SDK instance

You can initialize a Sharetribe SDK instance using only the
application's Client ID. Ensure the client ID is from a **Marketplace
API application** created in Console.

> ```jsx

Reviewer 3: ```suggestion
```js
```

Same goes for all the other code blocks too. I'm not sure if it has any effect on the code highlighting, maybe not.

### Initializing an Integration SDK instance

If you are accessing Integration API, you'll need to also provide the
client secret. Ensure the client ID is from an **Integration API
application** created in Console.

```jsx
const sharetribeIntegrationSdk = require('sharetribe-flex-integration-sdk');

const integrationSdk = sharetribeIntegrationSdk.createInstance({
  clientId: '<Client ID>',
  clientSecret: '<Client secret>',
});
```

<Callout type="warning">
  Keep your application's client secret secure. Never expose it to
  untrusted devices or applications such as end user browsers or mobile
  apps.
</Callout>

### Initializing the SDK with a trusted user access token

To make certain requests, you will need to access the Marketplace API
from a trusted context (i.e. on the server). In order to do this, you'll
need to log in using the SDK and perform a token exchange to retrieve a
trusted token.

<Steps>

{<h3>Create an instance of the SDK with the client ID, secret, and token
store</h3>}

```jsx
const sharetribeSdk = require('sharetribe-flex-sdk');

var sdk = sharetribeSdk.createInstance({
  clientId: '<Client ID>',
  clientSecret: '<Client secret>',

>   tokenStore: sharetribeSdk.tokenStore.memoryStore(),

Reviewer 3: I would remove this.

This depends a bit about the environment in which this piece of code is run. But since it's Marketplace API, we can assume that the user login happens in the marketplace frontend. The default store is cookie store, which allows the frontend and backend share the same access token.

</Callout>

```jsx
const sharetribeSdk = require('sharetribe-flex-sdk');

const transition = 'transition/request-payment';
const processAlias = 'default-booking/release-1';

const params = {
  listingId: '<Listing ID>',
  bookingStart: '2025-12-20T22:00:00.000Z',
  bookingEnd: '2025-12-21T22:00:00.000Z',
  lineItems: [
    {
      code: 'line-item/night',
      unitPrice: {
        _sdkType: 'Money',
        amount: 5000,
        currency: 'EUR',
      },
      quantity: 1,
      includeFor: ['customer', 'provider'],
    },
    {
      code: 'line-item/provider-commission',
      unitPrice: {
        _sdkType: 'Money',
        amount: 5000,
        currency: 'EUR',
      },
      percentage: -13,
      includeFor: ['provider'],
    },
  ],
};

const sdk = sharetribeSdk.createInstance({
  clientId: '<Client ID>',
  clientSecret: '<Client secret>',
});

sdk
  .login({ username: '<User email>', password: '<User password>' })
  .then((_) => {
    sdk.exchangeToken().then((res) => {
      var store = sharetribeSdk.tokenStore.memoryStore();
      store.setToken(res.data);

      var sdkTrusted = sharetribeSdk.createInstance({
        clientId: '<Client ID>',
        tokenStore: store,

>       });

Reviewer 3: This is slightly different than the one you used above where the token store creation was inline inside the SDK creation, i.e. `tokenStore: sharetribeSdk.tokenStore.memoryStore(res.data)`. Maybe use the same here so that it's easier to see the connection between the two?

  });
```

<Callout type="warning">
  Keep your application's client secret secure. Never expose it to
  untrusted devices or applications such as end user browsers or mobile
  apps.
</Callout>

### Transitioning a transaction

**API**: Marketplace API

This example shows how you can transition a transaction that has already
been initiated. The logged-in user must be a party that is allowed to
perform the transition - either the customer or provider. This can be
checked in the transaction process visualizer in Console, or from the
transaction's `process.edn` file.

<Callout type="info">
  Not all transitions require a trusted context. For non-privileged
  transitions, simply [logging
  in](https://www.sharetribe.com/docs/introduction/getting-started-with-sdks/#logging-in)
  is sufficient. For privileged transitions however, you'll need to use
  a trusted SDK instance. You can read more about privileged transitions
  [here](https://www.sharetribe.com/docs/concepts/transactions/privileged-transitions/).
</Callout>

```jsx
const sharetribeSdk = require('sharetribe-flex-sdk');

var sdk = sharetribeSdk.createInstance({
  clientId: '<Client ID>',
  clientSecret: '<Client secret>',

>   tokenStore: sharetribeSdk.tokenStore.memoryStore(),

Reviewer 3: Maybe remove this? Use the default cookie store
